cpp/src/ray/util/process_helper.cc

```diff
@@ -22,12 +25,44 @@ static std::string GetSessionDir(std::string redis_ip, int port, std::string pas
   return session_dir;
 }
 
-void ProcessHelper::StartRayNode(int redis_port, std::string redis_password,
-                                 int node_manager_port) {
-  std::vector<std::string> cmdargs(
-      {"ray", "start", "--head", "--port", std::to_string(redis_port), "--redis-password",
-       redis_password, "--node-manager-port", std::to_string(node_manager_port),
-       "--include-dashboard", "false"});
+/// IP address by which the local node can be reached *from* the `address`.
+///
+/// The behavior should be the same as `node_ip_address_from_perspective` from Ray Python
+/// code. See
+/// https://stackoverflow.com/questions/2674314/get-local-ip-address-using-boost-asio.
+///
+/// TODO(kfstorm): Make this function shared code and migrate Python & Java to use this
+/// function.
+///
+/// \param address The IP address and port of any known live service on the network
+/// you care about.
+/// \return The IP address by which the local node can be reached from the address.
+static std::string GetNodeIpAddress(const std::string &address = "8.8.8.8:53") {
+  std::vector<std::string> parts;
+  boost::split(parts, address, boost::is_any_of(":"));
+  RAY_CHECK(parts.size() == 2);
+  try {
+    boost::asio::io_service netService;
+    boost::asio::ip::udp::resolver resolver(netService);
+    boost::asio::ip::udp::resolver::query query(boost::asio::ip::udp::v4(), parts[0],
+                                                parts[1]);
+    boost::asio::ip::udp::resolver::iterator endpoints = resolver.resolve(query);
+    boost::asio::ip::udp::endpoint ep = *endpoints;
+    boost::asio::ip::udp::socket socket(netService);
+    socket.connect(ep);
```

Comment:
We'd better use a timeout here, and handle failure.

Reply:
The original Python code doesn't have a timeout either, and I already handled failure via try-catch.